runtime/common/src/apis.rs

```diff
@@ -345,23 +345,18 @@ macro_rules! impl_runtime_apis_plus_common {
 
 					// Because the staking solution calculates the next staking set at the beginning
 					// of the first block in the new round, the only way to accurately predict the
-					// authors would be to run the staking election while predicting. However this
-					// election is heavy and will take too long during prediction. So instead we
-					// work around it by always authoring the first slot in a new round. A longer-
-					// term solution will be to calculate the staking election result in the last
-					// block of the ending round.
-					if parachain_staking::Pallet::<Self>::round().should_update(block_number) {
-						log::info!(target: "nimbus-staking-workaround", "A new round is starting.\
-						Moonbeam will author during this slot without predicting eligibility first.\
-						You may see a `CannotBeAuthor` error soon. This is expected and harmless.\
-						It will be resolved soon.");
-
-						true
+					// authors is to run the staking election while predicting.
+					let mut round = parachain_staking::Pallet::<Self>::round();
+					if round.should_update(block_number) {
+						// mutate round to get next index
+						round.update(block_number);
+						// select top collator candidates for next round
+						parachain_staking::Pallet::<Self>::select_top_candidates(round.current);
```

Comment:
Looking back at this conversation, I still agree with your points about exposing this function that will mutate storage to arbitrary other pallets. But I'm not so convinced that only getting the election result without writing it to storage will work well for us.

Basically, I want to call `on_initialize` here, and it will always be public, and could be abused by other pallets in the runtime. The only reason I didn't literally call `on_initialize` was to skip the weight of making payouts.